README.md
# changecharitycomponents

Components for Change Charity flutter application

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [0.0.8] - small fixes

## [0.0.7] - added themes, and dark mode

## [0.0.6] - added autofill hints

## [0.0.5] - added null assertions, ability to change submit button color

## [0.0.4] - added example

## [0.0.3] - Paintings, SubmitRow, and TextField all imported through change_charity_components.dart

test/changecharitycomponents_test.dart
//import 'package:flutter_test/flutter_test.dart';
//
//import 'package:changecharitycomponents/text_field.dart';
//
//void main() {
//  test('adds one to input values', () {
//    final calculator = Calculator();
//    expect(calculator.addOne(2), 3);
//    expect(calculator.addOne(-7), -6);
//    expect(calculator.addOne(0), 1);
//    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//  });
//}


lib/theme.dart
import 'package:flutter/material.dart';

class ChangeTheme{
  ChangeTheme({this.isDark: false});
  final bool isDark;

  ThemeData get themeData {
   return isDark ? _darkTheme() : _lightTheme();
  }
}

ThemeData _lightTheme() {
  final ThemeData base = ThemeData.light();
  return base.copyWith(
    accentColor: Colors.lightBlue[200],
    primaryColor: Colors.lightBlueAccent[400],
    scaffoldBackgroundColor: Colors.grey[100],
    cardColor: Colors.blueGrey[200],
    textSelectionColor: Colors.lightBlueAccent[100],
    errorColor: Colors.redAccent,
    inputDecorationTheme: InputDecorationTheme(
      floatingLabelBehavior: FloatingLabelBehavior.never,
      filled: false,
      border: OutlineInputBorder(
        borderRadius: BorderRadius.all(Radius.circular(30)),
      ),
      enabledBorder: OutlineInputBorder(
          borderSide: BorderSide(
            style: BorderStyle.none,
            color: Colors.grey[900],
            width: 1.0,
          )
      ),
      focusedBorder: OutlineInputBorder(
        borderSide: BorderSide(
          style: BorderStyle.none,
          color: Colors.lightBlueAccent[400],
          width: 2.0,
        ),
      ),
      errorStyle: TextStyle(
        color: Colors.red,
        fontWeight: FontWeight.bold,
      ),
      errorBorder: OutlineInputBorder(
        borderSide: BorderSide(
          style: BorderStyle.none,
          color: Colors.red,
          width: 0,
        ),
      ),
      focusedErrorBorder: OutlineInputBorder(
        borderSide: BorderSide(
          style: BorderStyle.none,
          color: Colors.red[600],
          width: 0,
        ),
      ),
    ),
    buttonTheme: base.buttonTheme.copyWith(
      buttonColor: Colors.grey[300],
      shape:RoundedRectangleBorder(
        borderRadius: BorderRadius.all(Radius.circular(40)),
      ),
      textTheme: ButtonTextTheme.normal,
    ),
    bottomSheetTheme: BottomSheetThemeData(
      backgroundColor: Colors.blueGrey[400],
    ),
    primaryIconTheme: base.iconTheme.copyWith(color: Colors.grey[100]),
    textTheme: _lightTextTheme(base.textTheme),
    primaryTextTheme: _lightTextTheme(base.primaryTextTheme),
    accentTextTheme: _lightTextTheme(base.accentTextTheme),
  );
}

TextTheme _lightTextTheme(TextTheme base) {
  return base.copyWith(
    headline5: base.headline5.copyWith(
      fontWeight: FontWeight.w500,
    ),
    headline6: base.headline6.copyWith(
        fontSize: 18.0
    ),
    caption: base.caption.copyWith(
      fontWeight: FontWeight.w400,
      fontSize: 12.0,
    ),
    bodyText1: base.bodyText1.copyWith(
      fontWeight: FontWeight.w500,
      fontSize: 16.0,
    ),
  ).apply(
    fontFamily: 'Montserrat',
    displayColor: Colors.black,
    bodyColor: Colors.black,
  );
}

ThemeData _darkTheme() {
  final ThemeData base = ThemeData.dark();
  return base.copyWith(
    accentColor: Colors.lightBlue[200],
    primaryColor: Colors.lightBlueAccent[400],
    scaffoldBackgroundColor: Colors.grey[100],
    cardColor: Colors.blueGrey[200],
    textSelectionColor: Colors.lightBlue[400],
    errorColor: Colors.redAccent,
    inputDecorationTheme: InputDecorationTheme(
      floatingLabelBehavior: FloatingLabelBehavior.never,
      filled: false,
      border: OutlineInputBorder(
        borderRadius: BorderRadius.all(Radius.circular(30)),
      ),
      enabledBorder: OutlineInputBorder(
          borderSide: BorderSide(
            style: BorderStyle.none,
            color: Colors.grey[900],
            width: 1.0,
          )
      ),
      focusedBorder: OutlineInputBorder(
        borderSide: BorderSide(
          style: BorderStyle.none,
          color: Colors.lightBlueAccent[400],
          width: 2.0,
        ),
      ),
      errorStyle: TextStyle(
        color: Colors.red,
        fontWeight: FontWeight.bold,
      ),
      errorBorder: OutlineInputBorder(
        borderSide: BorderSide(
          style: BorderStyle.none,
          color: Colors.red,
          width: 0,
        ),
      ),
      focusedErrorBorder: OutlineInputBorder(
        borderSide: BorderSide(
          style: BorderStyle.none,
          color: Colors.red[600],
          width: 0,
        ),
      ),
    ),
    buttonTheme: base.buttonTheme.copyWith(
      buttonColor: Colors.grey[300],
      shape:RoundedRectangleBorder(
        borderRadius: BorderRadius.all(Radius.circular(40)),
      ),
      textTheme: ButtonTextTheme.normal,
    ),
    bottomSheetTheme: BottomSheetThemeData(
      backgroundColor: Colors.blueGrey[400],
    ),
    primaryIconTheme: base.iconTheme.copyWith(color: Colors.grey[100]),
    textTheme: _darkTextTheme(base.textTheme),
    primaryTextTheme: _darkTextTheme(base.primaryTextTheme),
    accentTextTheme: _darkTextTheme(base.accentTextTheme),
  );
}

TextTheme _darkTextTheme(TextTheme base) {
  return base.copyWith(
    headline5: base.headline5.copyWith(
      fontWeight: FontWeight.w500,
    ),
    headline6: base.headline6.copyWith(
        fontSize: 18.0
    ),
    caption: base.caption.copyWith(
      fontWeight: FontWeight.w400,
      fontSize: 12.0,
    ),
    bodyText1: base.bodyText1.copyWith(
      fontWeight: FontWeight.w500,
      fontSize: 16.0,
    ),
  ).apply(
    fontFamily: 'Montserrat',
    displayColor: Colors.white,
    bodyColor: Colors.white,
  );
}

example/main.dart
import 'package:flutter/material.dart';
import 'package:changecharitycomponents/change_charity_components.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Change Components Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TextEditingController _textController;
  FocusNode _focusNode;
  String _errMsg = '';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: ChangeTextInput(
        autofillHint: AutofillHints.password,
        controller: _textController,
        focusNode: _focusNode,
        hintText: "Password",
        prefixIcon: Icons.lock,
        isPassword: true,
        errMsg: _errMsg,
        errFunc: (s) {
          setState(() {
            _errMsg = s;
          });
        },
      ),
    );
  }
}


lib/submit_row.dart
library submit_row;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

class ChangeSubmitRow extends StatefulWidget {
  ChangeSubmitRow(
      {@required this.animation,
      @required this.onClick,
      @required this.loading,
      this.text: '',
      this.scale: 1,
      this.margin,
      this.buttonColors: const [Color(0xFF29B6F6), Color(0xFF4FC3F7)],
      this.arrowColor: const Color(0xFFFFFFFF)})
      : assert(animation != null),
        assert(onClick != null),
        assert(loading != null);

  final Animation<Color> animation;
  final String text;
  final double scale;
  final Function onClick;
  final bool loading;
  final List<Color> buttonColors;
  final Color arrowColor;
  final EdgeInsets margin;

  @override
  _ChangeSubmitRowState createState() => _ChangeSubmitRowState();
}

class _ChangeSubmitRowState extends State<ChangeSubmitRow> {
  Widget _text() {
    return Container(
      margin: EdgeInsets.fromLTRB(10, 10, 10, 0),
      child: Text(
        widget.text,
        style: TextStyle(
          fontSize: 33,
        ),
      ),
    );
  }

  Widget _button() {
    if (widget.loading) {
      return Container(
        margin: EdgeInsets.fromLTRB(22, 0, 32, 0),
        child: CircularProgressIndicator(
          valueColor: widget.animation,
        ),
      );
    } else {
      return Container(
        margin: EdgeInsets.fromLTRB(0, 10, 20, 0),
        width: 70,
        child: RaisedButton(
          onPressed: () {
            FocusScope.of(context).unfocus();
            widget.onClick();
          },
          padding: EdgeInsets.symmetric(vertical: 0, horizontal: 0),
          elevation: 10,
          shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.all(Radius.circular(60))),
          child: Ink(
            width: 70,
            height: 40,
            decoration: BoxDecoration(
                gradient: LinearGradient(
                  colors: widget.buttonColors,
                  begin: Alignment.topCenter,
                  end: Alignment.bottomCenter,
                ),
                borderRadius: BorderRadius.circular(30.0)),
            child: Icon(
              Icons.arrow_forward,
              color: widget.arrowColor,
              size: 27,
            ),
          ),
        ),
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        return Transform.scale(
          scale: widget.scale,
          child: Align(
            alignment: Alignment.centerRight,
            child: Container(
              margin: widget.margin,
              width: MediaQuery.of(context).size.width,
              height: MediaQuery.of(context).size.height > 700 ? 70 : 60,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.end,
                children: <Widget>[
                  _text(),
                  _button(),
                ],
              ),
            ),
          ),
        );
      },
    );
  }
}


lib/text_field.dart
library text_field;

import 'package:flutter/material.dart';

class ChangeTextInput extends StatefulWidget {
  ChangeTextInput({
    @required this.controller,
    @required this.focusNode,
    @required this.hintText,
    @required this.prefixIcon,
    @required this.autofillHint,
    this.isPassword: false,
    this.last: false,
    this.errFunc,
    this.errMsg: '',
    this.lastFunc,
  })  : assert(controller != null),
        assert(focusNode != null),
        assert(hintText != null),
        assert(prefixIcon != null),
        assert(autofillHint != null);

  final TextEditingController controller;
  final FocusNode focusNode;
  final IconData prefixIcon;
  final String hintText;
  final String errMsg;
  final bool isPassword;
  final bool last;
  final Function errFunc;
  final Function lastFunc;
  final String autofillHint;

  @override
  _ChangeTextInputState createState() => _ChangeTextInputState();
}

class _ChangeTextInputState extends State<ChangeTextInput> {
  bool _obscurePass = true;

  Widget _textContainer(constraints) {
    return Container(
      margin: EdgeInsets.only(
          right: constraints.maxWidth * 0.05,
          left: constraints.maxWidth * 0.05),
      decoration: BoxDecoration(
        color: MediaQuery.of(context).platformBrightness == Brightness.light ? Colors.white : Colors.black,
        borderRadius: BorderRadius.all(Radius.circular(30)),
        boxShadow: [
          BoxShadow(
            color: MediaQuery.of(context).platformBrightness == Brightness.light ? Colors.grey[350] : Colors.grey[600],
            blurRadius: 20.0,
            offset: Offset.fromDirection(0.9),
          ),
        ],
      ),
      child: TextField(
        autofillHints: [widget.autofillHint],
        controller: widget.controller,
        obscureText: widget.isPassword ? _obscurePass : false,
        onChanged: (s) {
          setState(() {
            widget.errFunc("");
          });
        },
        onEditingComplete: () {
          widget.last
              ? widget.focusNode.unfocus()
              : widget.focusNode.nextFocus();
          widget.lastFunc();
        },
        decoration: InputDecoration(
          labelText: widget.hintText,
          floatingLabelBehavior: FloatingLabelBehavior.never,
          prefixIcon: Container(
            margin: EdgeInsets.only(left: 25, right: 15),
            child: Icon(
              widget.prefixIcon,
              size: 20,
              color: MediaQuery.of(context).platformBrightness == Brightness.light ? Colors.black : Colors.white,
            ),
          ),
          suffixIcon: widget.isPassword ? _passSuffix() : null,
        ),
        focusNode: widget.focusNode,
        textInputAction:
            widget.last ? TextInputAction.done : TextInputAction.next,
      ),
    );
  }

  Widget _errContainer() {
    return Align(
      alignment: Alignment.centerLeft,
      child: Container(
        margin: EdgeInsets.only(
          left: 40,
          top: 2,
          bottom: 3,
        ),
        height: 25,
        child: Text(
          widget.errMsg,
          style: TextStyle(
            color: Colors.red,
            fontWeight: FontWeight.bold,
          ),
        ),
      ),
    );
  }

  Widget _passSuffix() {
    return Container(
      margin: EdgeInsets.only(left: 15, right: 20),
      child: IconButton(
        splashColor: Colors.transparent,
        onPressed: () {
          setState(() {
            _obscurePass = !_obscurePass;
          });
        },
        icon: Icon(
          _obscurePass ? Icons.visibility : Icons.visibility_off,
          size: 20,
          color: MediaQuery.of(context).platformBrightness == Brightness.light ? Colors.black : Colors.white,
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        return Column(
          children: <Widget>[
            _textContainer(constraints),
            _errContainer(),
          ],
        );
      },
    );
  }
}


lib/paintings.dart
library paintings;

import 'package:flutter/material.dart';
import 'package:flutter/painting.dart';

class ChangeBankPaint extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    var paintA = Paint()
      ..color = Colors.lightBlue[200]
      ..style = PaintingStyle.fill;
    var paintB = Paint()
      ..color = Colors.lightBlue[400].withOpacity(0.8)
      ..style = PaintingStyle.fill;

    var pathA = Path();
    pathA.moveTo(size.width * 0.6, -100);
    pathA.quadraticBezierTo(size.width * 0.75, size.height * 0.14,
        size.width * 0.65, size.height * 0.2);
    pathA.quadraticBezierTo(size.width * 0.5, size.height * 0.28,
        size.width * 0.25, size.height * 0.2);
    pathA.quadraticBezierTo(
        size.width * 0.15, size.height * 0.17, 0, size.height * 0.25);
    pathA.lineTo(0, -100);
    canvas.drawPath(pathA, paintA);

    var pathB = Path();
    pathB.moveTo(0, size.height * 0.28);
    pathB.quadraticBezierTo(size.width * 0.18, size.height * 0.14,
        size.width * 0.33, size.height * 0.15);
    pathB.quadraticBezierTo(size.width * 0.62, size.height * 0.23,
        size.width * 0.58, size.height * 0.1);
    pathB.quadraticBezierTo(
        size.width * 0.4, -200, size.width * 0.25, size.height * 0.01);
    pathB.quadraticBezierTo(
        size.width * 0.2, size.height * 0.1, 0, size.height * 0.03);
    canvas.drawPath(pathB, paintB);
  }

  @override
  bool shouldRepaint(ChangeBankPaint oldDelegate) => false;
}

class ChangeLoginPaint extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    var paintA = Paint()
      ..color = Colors.lightBlue[200]
      ..style = PaintingStyle.fill;
    var paintB = Paint()
      ..color = Colors.lightBlue[400].withOpacity(0.8)
      ..style = PaintingStyle.fill;

    var pathA = Path();
    pathA.moveTo(size.width, -70);
    pathA.lineTo(size.width * 0.75, -75);
    pathA.quadraticBezierTo(size.width * 0.37, size.height * 0.1,
        size.width * 0.63, size.height * 0.13);
    pathA.quadraticBezierTo(size.width * 0.83, size.height * 0.15,
        size.width * 0.8, size.height * 0.25);
    pathA.quadraticBezierTo(
        size.width * 0.8, size.height * 0.35, size.width, size.height * 0.26);
    canvas.drawPath(pathA, paintA);

    var pathB = Path();
    pathB.moveTo(size.width, -70);
    pathB.lineTo(size.width * 0.6, -70);
    pathB.quadraticBezierTo(size.width * 0.53, size.height * 0.1,
        size.width * 0.7, size.height * 0.1);
    pathB.quadraticBezierTo(size.width * 0.9, size.height * 0.11,
        size.width * 0.82, size.height * 0.2);
    pathB.quadraticBezierTo(
        size.width * 0.67, size.height * 0.33, size.width, size.height * 0.25);
    canvas.drawPath(pathB, paintB);
  }

  @override
  bool shouldRepaint(ChangeLoginPaint oldDelegate) => false;
}

class ChangeSignUpPaint extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    var paintA = Paint()
      ..color = Colors.lightBlue[200]
      ..style = PaintingStyle.fill;
    var paintB = Paint()
      ..color = Colors.lightBlue[400].withOpacity(0.8)
      ..style = PaintingStyle.fill;

    var pathA = Path();
    pathA.moveTo(0, size.height * 0.35);
    pathA.quadraticBezierTo(size.width * 0.5, size.height * 0.32,
        size.width * 0.38, size.height * 0.42);
    pathA.quadraticBezierTo(size.width * 0.12, size.height * 0.6,
        size.width * 0.35, size.height * 0.7);
    pathA.quadraticBezierTo(
        size.width * 0.6, size.height * 0.78, 0, size.height * 0.85);
    canvas.drawPath(pathA, paintA);

    var pathB = Path();
    pathB.moveTo(0, size.height * 0.3);
    pathB.quadraticBezierTo(size.width * 0.39, size.height * 0.25,
        size.width * 0.25, size.height * 0.42);
    pathB.quadraticBezierTo(size.width * 0.05, size.height * 0.6,
        size.width * 0.3, size.height * 0.62);
    pathB.quadraticBezierTo(
        size.width * 0.52, size.height * 0.65, 0, size.height * 0.75);
    canvas.drawPath(pathB, paintB);
  }

  @override
  bool shouldRepaint(ChangeSignUpPaint oldDelegate) => false;
}

class ChangeSearchPaint extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    var paintA = Paint()
      ..color = Colors.lightBlue[200]
      ..style = PaintingStyle.fill;
    var paintB = Paint()
      ..color = Colors.lightBlue[400].withOpacity(0.8)
      ..style = PaintingStyle.fill;

    var pathA = Path();
    pathA.moveTo(size.width + 100, size.height * 0.8);
    pathA.quadraticBezierTo(size.width * 0.7, size.height * 0.72,
        size.width * 0.68, size.height * 0.85);
    pathA.quadraticBezierTo(size.width * 0.68, size.height * 0.93,
        size.width * 0.52, size.height * 0.93);
    pathA.quadraticBezierTo(
        size.width * 0.43, size.height * 0.93, size.width * 0.4, size.height);
    pathA.lineTo(size.width + 100, size.height + 100);
    canvas.drawPath(pathA, paintA);

    var pathB = Path();
    pathB.moveTo(size.width + 100, size.height * 0.88);
    pathB.quadraticBezierTo(size.width * 0.89, size.height * 0.6,
        size.width * 0.82, size.height * 0.8);
    pathB.quadraticBezierTo(size.width * 0.8, size.height * 0.88,
        size.width * 0.58, size.height * 0.91);
    pathB.quadraticBezierTo(
        size.width * 0.49, size.height * 0.93, size.width * 0.48, size.height);
    pathB.lineTo(size.width + 100, size.height + 100);
    canvas.drawPath(pathB, paintB);
  }

  @override
  bool shouldRepaint(ChangeSearchPaint oldDelegate) => false;
}

class ChangeHomePaint extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    var paintA = Paint()
      ..color = Colors.lightBlue[200]
      ..style = PaintingStyle.fill;
    var paintB = Paint()
      ..color = Colors.lightBlue[400].withOpacity(.8)
      ..style = PaintingStyle.fill;
    var pathA = Path();
    pathA.moveTo(size.width, size.height * .25);
    pathA.lineTo(size.width, size.height * .65);
    pathA.quadraticBezierTo(size.width - 200, size.height * .55,
        size.width - 100, size.height * .45);
    pathA.quadraticBezierTo(
        size.width, size.height * .35, size.width - 150, size.height * .3);
    pathA.quadraticBezierTo(
        size.width - 250, size.height * .25, size.width, size.height * .2);
    var pathB = Path();
    pathB.moveTo(size.width, size.height * .15);
    pathB.lineTo(size.width, size.height * .7);
    pathB.quadraticBezierTo(size.width - 200, size.height * .55,
        size.width - 100, size.height * .48);
    pathB.quadraticBezierTo(size.width + 100, size.height * .35,
        size.width - 100, size.height * .35);
    pathB.quadraticBezierTo(
        size.width - 350, size.height * .35, size.width, size.height * .25);

    canvas.drawPath(pathA, paintA);
    canvas.drawPath(pathB, paintB);
  }

  @override
  bool shouldRepaint(ChangeHomePaint oldDelegate) => false;
}

class ChangeProfilePaint extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    var paintA = Paint()
      ..color = Colors.lightBlue[200]
      ..style = PaintingStyle.fill;
    var paintB = Paint()
      ..color = Colors.lightBlue[400].withOpacity(0.8)
      ..style = PaintingStyle.fill;

    var pathA = Path();
    pathA.moveTo(size.width, size.height + 50);
    pathA.lineTo(size.width, size.height * .85);
    pathA.quadraticBezierTo(
        size.width * .8, size.height * .78, size.width * .72, size.height * .85);
    pathA.quadraticBezierTo(
        size.width * .6, size.height * .95, size.width * .3, size.height * .9);
    pathA.quadraticBezierTo(
        size.width * .1, size.height * .88, size.width * .2, size.height + 50);
    canvas.drawPath(pathA, paintA);

    var pathB = Path();
    pathB.moveTo(size.width, size.height + 50);
    pathB.lineTo(size.width, size.height * .95);
    pathB.quadraticBezierTo(
        size.width * .8, size.height * .7, size.width * .55, size.height * .9);
    pathB.quadraticBezierTo(
        size.width * .45, size.height, size.width * .2, size.height * .9);
    pathB.quadraticBezierTo(
        size.width * .05, size.height * .85, size.width * .1, size.height + 50);
    canvas.drawPath(pathB, paintB);
  }

  @override
  bool shouldRepaint(ChangeProfilePaint oldDelegate) => false;
}
